### PR TITLE
 [Fix] Database owner login is unable to view db objects in SSMS

### PR DESCRIPTION
**Pull request by aswiniip**
### Description

Currently in babelfish when we create a login (e.g. postjend_svc) and a database (e.g. test) and make postjend_svc owner of test and then connect to SSMS object explorer as login "postjend_svc" to database "test" and try to expand the "Database" field to view the objects. It does not show anything. With this change the login with be able to expand the "Database" field.

The issue arised because when the databases are expanded in SSMS, a query is being executed which contains a cross-db query. Currently, we handle cross-db queries by switching the database and current user. This cross-db query is resulting into permission denied error due to lack of permission of the current user (which is switched internally to master_guest) to insert into a table created in "test" db.

The solution of the above issue is implemented in three parts:

1 - For the successful execution of the cross-db query, I switched the current user with session user (login) which resulted in the user having adequate permission of both the databases required for the execution of the script.

BABEL-5218 was found after implementing the above fix. This issue arised because when a non sysadmin login tries to run cross-db dml query like INSERT. The permission check is done through the pg_catalogue side causing a permission denied error because dbo role for that database is not given to that login during the "alter authorization".

2 - To fix the above issue, I assigned dbo role to the login when it is made the owner of the database. Since the login has the dbo role, it can perform the above queries successfully.

After implementing the above fixes, an issue was found with the `has_dbaccess()` function which returns incorrect value for non-sysadmin logins who are db_owners. The logins who are db_owner should get the output of `has_dbaccess()` as `1`.

3 - To fix the above issue, I added a check in `has_dbaccess()` function which checks if the login is a db_owner. By doing so the function returns `1` for users who are db_owner and `0` otherwise.

### Issues Resolved

BABEL-5119, BABEL-5218

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

Signed-off-by: P Aswini Kumar [aswinii@amazon.com](mailto:aswinii@amazon.com)

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pl_exec.c [hunk 1]
```diff
@@ -4661,7 +4661,10 @@ exec_stmt_execsql(PLtsql_execstate *estate,
 		 * schemas, Change the session property.
 		 */
 		if (stmt->schema_name != NULL && (strcmp(stmt->schema_name, "sys") == 0 || strcmp(stmt->schema_name, "information_schema") == 0))
+		{
 			set_session_properties(stmt->db_name);
+			SetCurrentRoleId(GetSessionUserId(), false);
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/pl_exec.c:4666]: Why is this change needed?

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]
Upgrade changes?

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 2]
```diff
@@ -2855,7 +2858,7 @@ has_dbaccess(PG_FUNCTION_ARGS)
 		Oid			datdba;
 
 		datdba = get_role_oid("sysadmin", false);
-		if (is_member_of_role(GetSessionUserId(), datdba))
+		if (is_member_of_role(GetSessionUserId(), datdba) || login_is_db_owner)
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/runtime/functions.c:2864]: please add the comment why have you added login_is_db_owner in the check.

**Review by shalinilohia50 — CHANGES_REQUESTED** [review]
You would need to write tests in vu-prepare, vu-verify, vu-cleanup format. Also, add tests in schedule file of upgrade.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 3]
```diff
@@ -3906,9 +3909,29 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
...
+
+	/* Save the previous user to be restored after granting dbo role to the login. */
+	GetUserIdAndSecContext(&save_userid, &save_sec_context);
+
+	PG_TRY();
+	{
+		/*
+		* Set current user to bbf_role_admin for grant roles.
+		*/
+		SetUserIdAndSecContext(get_bbf_role_admin_oid(), save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:3921]: What if we switch to login(i.e. SessionUser instead)? That will be more secure since bbf_role_admin will have all the privileges but session user will have privileges corresponding only to set of users mapped to it.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 4]
```diff
@@ -1083,6 +1083,19 @@ update_GrantRoleStmt(Node *n, List *privs, List *roles)
 	stmt->grantee_roles = roles;
 }
 
+void
+update_RevokeRoleStmt(Node *n, List *privs, List *roles)
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Let's reuse update_GrantRoleStmt. We don't need to mark `is_grant = false`, it will be done implicitly while parsing the statement.

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 5]
```diff
@@ -585,6 +585,86 @@ grant_guests_to_login(const char *login)
...
+	wrapper = makeNode(PlannedStmt);
+	wrapper->commandType = CMD_UTILITY;
+	wrapper->canSetTag = false;
+	wrapper->utilityStmt = stmt;
+	wrapper->stmt_location = 0;
+	wrapper->stmt_len = 18;
+
+	/* do this step */
+	ProcessUtility(wrapper,
+				   "(CREATE DATABASE )",
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Let's keep it something else "(ALTER DATABASE OWNER )" and accordingly let's update the stmt_len.

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_exec-2.c (hunk 3 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3921]: As discussed offline, we are sticking with the `bbf_role_admin` role to grant roles

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 4 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 5 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done

File: test/JDBC/expected/BABEL-5119-vu-verify.out [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+SELECT name FROM master.sys.databases WHERE name LIKE 'BABEL5119%'
+ORDER BY name
+GO
+~~START~~
+varchar
+babel5119_1
+~~END~~
+
+
+SELECT name FROM master.sys.views WHERE name LIKE 'BABEL5119%'
```
Comment (anju15bharti) [inline test/JDBC/expected/BABEL-5119-vu-verify.out:166]: What are we trying to check with this views output?
Please write one line comment for each cases that what are we trying to verify here.

File: test/JDBC/expected/BABEL-5119-vu-verify.out [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+CREATE LOGIN login_babel5119_1 WITH PASSWORD = '12345678'
+CREATE LOGIN login_babel5119_2 WITH PASSWORD = '12345678'
+ALTER authorization on database::BABEL5119_1 TO login_babel5119_1
+GO
+-- terminate-tsql-conn
+
+-- tsql user=login_babel5119_1 password=12345678
+USE master
+GO
+CREATE TABLE BABEL5119_t4(a int)
```
Comment (anju15bharti) [inline test/JDBC/expected/BABEL-5119-vu-verify.out:12]: why this login has privilege to create objects in master database?
Also do print current_user here.

File: test/JDBC/expected/BABEL-5119-vu-verify.out [hunk 8]
```diff
@@ -0,0 +1,186 @@
+-- tsql
+CREATE LOGIN login_babel5119_1 WITH PASSWORD = '12345678'
+CREATE LOGIN login_babel5119_2 WITH PASSWORD = '12345678'
+ALTER authorization on database::BABEL5119_1 TO login_babel5119_1
```
Comment (anju15bharti) [inline test/JDBC/expected/BABEL-5119-vu-verify.out]: Add db or something similar in database name to make it recognizable.

File: test/JDBC/expected/BABEL-5119-vu-verify.out [hunk 9]
```diff
@@ -0,0 +1,186 @@
+-- tsql
+CREATE LOGIN login_babel5119_1 WITH PASSWORD = '12345678'
+CREATE LOGIN login_babel5119_2 WITH PASSWORD = '12345678'
```
Comment (anju15bharti) [inline test/JDBC/expected/BABEL-5119-vu-verify.out:3]: why login creation in this file?

File: test/JDBC/expected/BABEL-5119-vu-prepare.out [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+USE BABEL5119_1
+GO
+CREATE TABLE BABEL5119_t1(a int)
+GO
+CREATE TABLE BABEL5119_t2(a int)
+GO
+CREATE TABLE BABEL5119_t3(a int)
+GO
+INSERT INTO BABEL5119_t1 VALUES(1)
```
Comment (anju15bharti) [inline test/JDBC/expected/BABEL-5119-vu-prepare.out]: Insertion should be performed in verify file not create until we don't use the inserted value in this script itself.

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 11]
```diff
@@ -585,6 +585,79 @@ grant_guests_to_login(const char *login)
 	pfree(query.data);
 }
 
+void
+grant_dbo_to_login(const char* login, const char* db_name, bool is_grant)
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Since we are using the same API to grant and revoke dbo, it's name should reflect the same. Lets rename it to something like grant_revoke_dbo_to_login
Please add comment for this API what actually it does.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 12]
```diff
@@ -28,6 +28,7 @@
 #include "parser/scansup.h"
 #include "parser/parse_oper.h"
 #include "src/include/lib/qunique.h"
+#include "rolecmds.h"
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: +1

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 13]
```diff
@@ -70,6 +70,7 @@
 #include "catalog/pg_trigger.h"
 #include "catalog/pg_constraint.h"
 #include "parser/parse_oper.h"
+#include "../src/dbcmds.h"
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/runtime/functions.c]: the header should be in lexicographical order

**Review by anju15bharti — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/BABEL-5119-vu-verify.out (hunk 7 above)
Comment (aswiniip) [reply test/JDBC/expected/BABEL-5119-vu-verify.out:12]: It creates the objects in guest schema of the master database.

File: test/JDBC/expected/BABEL-5119-vu-verify.out (hunk 9 above)
Comment (aswiniip) [reply test/JDBC/expected/BABEL-5119-vu-verify.out:3]: Currently in babelfish, during upgrades logins are dumped without passwords so it is required to reset their passwords on the target server otherwise authentication will fail for those logins. So I created the logins in the verify itself.

File: test/JDBC/input/BABEL-5119-vu-prepare.mix [hunk 14]
```diff
@@ -0,0 +1,27 @@
+-- tsql
+CREATE DATABASE BABEL5119_1
+GO
+
+USE BABEL5119_1
+GO
```
Comment (HarshLunagariya) [inline test/JDBC/input/BABEL-5119-vu-prepare.mix]: Current framework runs on multi-db , right? If yes, then can we please create few more user-created database which are owned by non-sysadmin users and then let's execute has_dbaccess AND some basic permission check. Then only our upgrade changes will be verified.

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 2 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/runtime/functions.c:2864]: Added a comment

File: test/JDBC/input/BABEL-5119-vu-prepare.mix (hunk 14 above)
Comment (aswiniip) [reply test/JDBC/input/BABEL-5119-vu-prepare.mix]: Creating multiple databases resulted in failure of upgrade tests because the upgrade tests are run in single db mode

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 13 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Done

File: contrib/babelfishpg_tsql/src/pl_exec-2.c (hunk 12 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 11 above)
Comment (aswiniip) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done and added a comment with brief description.

File: test/JDBC/expected/BABEL-5119-vu-prepare.out (hunk 10 above)
Comment (aswiniip) [reply test/JDBC/expected/BABEL-5119-vu-prepare.out]: Removed insert query from the vu-prepare file

File: test/JDBC/expected/BABEL-5119-vu-verify.out (hunk 8 above)
Comment (aswiniip) [reply test/JDBC/expected/BABEL-5119-vu-verify.out]: Changed `BABEL5119_1` to `BABEL5119_db`

File: test/JDBC/expected/BABEL-5119-vu-verify.out (hunk 7 above)
Comment (aswiniip) [reply test/JDBC/expected/BABEL-5119-vu-verify.out:12]: Added a comment for better clarity in future

**Review by HarshLunagariya — APPROVED** [review]

**Review by shalinilohia50 — APPROVED** [review]
Looks good to me.

**Review by anju15bharti — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
